types/prop-types/prop-types-tests.ts

```diff
@@ -1,29 +1,161 @@
+import { ReactElement, ReactNode } from "react";
 import * as PropTypes from "prop-types";
 
+declare const uniqueType: unique symbol;
+
+class TestClass { }
+
 interface Props {
-    any: any;
+    any?: any;
     array: string[];
     bool: boolean;
-    func: any;
-    string: string;
+    element: ReactElement<any>;
+    func(foo: string): void;
+    node?: ReactNode;
+    requiredNode: NonNullable<ReactNode>;
     number: number;
+    object: object;
+    string: string;
     symbol: symbol;
-    object: {};
-    node: any;
-    element: any;
+    instanceOf: TestClass;
+    oneOf: 'a' | 'b' | 'c';
+    oneOfType: string | boolean | {
+        foo?: string;
+        bar: number;
+    };
+    numberOrFalse: false | number;
+    nodeOrRenderFn?: ReactNode | (() => ReactNode);
+    arrayOf: boolean[];
+    objectOf: { [K: string]: number };
+    shape: {
+        foo: string;
+        bar?: boolean;
+        baz?: any
+    };
+    optionalNumber?: number | null;
+    customProp?: typeof uniqueType;
 }
 
-const propTypes: PropTypes.ValidationMap<Props> = {
-    any: PropTypes.any.isRequired,
+const innerProps = {
+    foo: PropTypes.string.isRequired,
+    bar: PropTypes.bool,
+    baz: PropTypes.any
+};
+
+const arrayOfTypes = [PropTypes.string, PropTypes.bool, PropTypes.shape({
+    foo: PropTypes.string,
+    bar: PropTypes.number.isRequired
+})];
+type PropTypesMap = PropTypes.ValidationMap<Props>;
+
+// TS checking
+const propTypes: PropTypesMap = {
+    any: PropTypes.any,
     array: PropTypes.array.isRequired,
     bool: PropTypes.bool.isRequired,
+    element: PropTypes.element.isRequired,
     func: PropTypes.func.isRequired,
+    node: PropTypes.node,
+    requiredNode: PropTypes.node.isRequired,
     number: PropTypes.number.isRequired,
     object: PropTypes.object.isRequired,
     string: PropTypes.string.isRequired,
     symbol: PropTypes.symbol.isRequired,
-    node: PropTypes.node.isRequired,
-    element: PropTypes.element.isRequired
+    instanceOf: PropTypes.instanceOf(TestClass).isRequired,
+    oneOf: PropTypes.oneOf(['a', 'b', 'c']).isRequired,
```

Comment:
Yeah, I'd check this independently of that type annotation on the variable declaration (as that may cause this position to be literally typed independently of the constraint on the generic parameter)

Reply:
In order to get the types to match on the un-annotated prop types, I had to specify the generic parameter (see line 95). Couldn't stop the compiler from widening the type :s